### PR TITLE
applications: asset_tracker: Add motion module

### PR DESCRIPTION
**Pull request by jhn-nordic**
Refactor existing flip detection functionality to a motion module
and add encoding of motion data to cloud.

Signed-off-by: Jon Helge Nistad <jon.helge.nistad@nordicsemi.no>

### CODE REVIEW

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 1]
```diff
@@ -458,6 +458,32 @@ int cloud_encode_env_sensors_data(const env_sensor_data_t *sensor_data,
 	return cloud_encode_data(&cloud_sensor, output);
 }
 
+int cloud_encode_motion_data(const motion_data_t *motion_data,
+				  struct cloud_msg *output)
+{
+	__ASSERT_NO_MSG(motion_data != NULL);
+	__ASSERT_NO_MSG(output != NULL);
+
+	struct cloud_channel_data cloud_sensor;
```
Comment (SebastianBoe) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: Does/could/should there exist an init function for cloud_channel_data ?

As this is written, if another field is added in the future, and this field is read by cloud_encode_data, then we might accidentally forget to initialize that field from this function and get undefined behaviour when cloud_encode_data reads it.

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 2]
```diff
@@ -458,6 +458,32 @@ int cloud_encode_env_sensors_data(const env_sensor_data_t *sensor_data,
...
+	cloud_sensor.type = CLOUD_CHANNEL_FLIP;
+
+	switch (motion_data->orientation) {
+	case MOTION_ORIENTATION_NORMAL:
+		cloud_sensor.data.buf = "NORMAL";
+		cloud_sensor.data.len = sizeof("NORMAL") - 1;
+		break;
+	case MOTION_ORIENTATION_UPSIDE_DOWN:
+		cloud_sensor.data.buf = "UPSIDE_DOWN";
+		cloud_sensor.data.len = sizeof("UPSIDE_DOWN") - 1;
```
Comment (SebastianBoe) [inline applications/asset_tracker/src/cloud_codec/cloud_codec.c]: char * buf;

```
switch([...]) {
buf = "NORMAL";
[...]
buf = "UPSIDE_DOWN";
[...]
}
cloud_sensor.data.buf = buf;
cloud_sensor.data.len = sizeof(buf) - 1;
```

File: applications/asset_tracker/src/motion/motion.c [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+#include <stdio.h>
+#include <string.h>
+#include <sensor.h>
+#include "motion.h"
+
+motion_handler_t handler;
+static struct device *accel_dev;
+
+#define FLIP_ACCELERATION_THRESHOLD	5.0
+#define ACCEL_INVERTED			CONFIG_ACCEL_INVERTED
```
Comment (SebastianBoe) [inline applications/asset_tracker/src/motion/motion.c]: Unnecessary indirection IMHO.

File: applications/asset_tracker/src/motion/motion.c [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	if (orientation == NULL || acceleration_data == NULL) {
+		return -EINVAL;
+	}
+
+	if (acceleration_data->z >= FLIP_ACCELERATION_THRESHOLD) {
+		if (IS_ENABLED(ACCEL_INVERTED)) {
+			*orientation = MOTION_ORIENTATION_UPSIDE_DOWN;
+		} else {
+			*orientation = MOTION_ORIENTATION_NORMAL;
+		}
```
Comment (SebastianBoe) [inline applications/asset_tracker/src/motion/motion.c]:
```
*orientation = IS_ENABLED(ACCEL_INVERTED) ?               \
                         MOTION_ORIENTATION_UPSIDE_DOWN : \
                         MOTION_ORIENTATION_NORMAL;
```

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c (hunk 2 above)
Comment (jhn-nordic) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: done

File: applications/asset_tracker/src/motion/motion.c (hunk 3 above)
Comment (jhn-nordic) [reply applications/asset_tracker/src/motion/motion.c]: fixed

File: applications/asset_tracker/src/motion/motion.c (hunk 4 above)
Comment (jhn-nordic) [reply applications/asset_tracker/src/motion/motion.c]: fixed

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c (hunk 1 above)
Comment (jhn-nordic) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: I see your point. We will add this as a task to our backlog of improvements

**Comment by jhn-nordic** [comment]
@SebastianBoe Thanks for your review. I have just pushed fixes that should address your findings.

**Pull request merged**